src/js_native_api.h

```diff
@@ -102,6 +102,34 @@ node_api_symbol_for(napi_env env,
                     size_t length,
                     napi_value* result);
 #endif  // NAPI_EXPERIMENTAL
+
+#ifdef NAPI_EMBEDDING
+NAPI_EXTERN napi_status NAPI_CDECL napi_create_platform(int argc,
+                                                        char** argv,
+                                                        int exec_argc,
+                                                        char** exec_argv,
+                                                        char*** errors,
+                                                        int thread_pool_size,
+                                                        napi_platform* result);
+
+NAPI_EXTERN napi_status NAPI_CDECL
+napi_destroy_platform(napi_platform platform);
+
+NAPI_EXTERN napi_status NAPI_CDECL
+napi_create_environment(napi_platform platform,
```

Comment:
Please note that the term `Environment` is somehow overused in Node.JS. 
The `napi_env` is different from the Node.JS `Environment`. 
We create one instance of `napi_env` per each native module. It allows using the `napi_env` instance settings and other flags per module. Each Node.JS environment may have zero or more `napi_env` instances depending on how many Node-API modules it loads.
It is important to clarify what type of environment the `napi_create_environment` function creates. I bet it is not the `napi_env`, but rather a moniker to the Node.JS `Envrironment`, which I guess must have a different type such as `napi_node_instance`, `napi_node_env`, or something else along those lines.

Reply:
`napi_create_environment` creates a `napi_env` instance linked to a Node `Environment` containing an additional context pointer (`set_embedded`/`get_embedded`). In theory (one must take care to register the cleanup only once) there can also be multiple `napi_env` instances for a single embedded `Environment` - but I don't see any reason why.

The whole point of the NAPI embedding is that it allows reusing of the NAPI functions for the embedded environment. So whatever type I return, that type must remain compatible with all the NAPI functions. In the C++ API, I return a different type (inherited from the basic `Node::Env`), but I don't see any practical way to do this in C. If I return a different type, you will have to cast it every time you invoke a NAPI function.

I don't understand the problem with the current situation - every `napi_env` is linked to a single `Environment`. When embedding, it is the `Environment` that is slightly different, not the `napi_env`.

Reply:
If I name this function `napi_create_instance` but still return a `napi_env`, does this address your concern?

Reply:
Or there is the quite cumbersome option:
```c
napi_embedded_instance inst;
napi_create_instance(platform, NULL, NULL, &inst);
napi_value global;
napi_get_global(napi_instance_env(inst), &global);
```

It is ugly in C, but it will be covered up in C++